packages/server/src/server.ts

```diff
@@ -1,445 +1,72 @@
-
-            // for response logging
-            res.locals.didRenderBanner = !!response.data;
-            res.locals.clientName = tracking.clientName;
-            // be specific about which fields to log, to avoid accidentally logging inappropriate things in future
-            res.locals.bannerTargeting = {
-                shouldHideReaderRevenue: targeting.shouldHideReaderRevenue,
-                showSupportMessaging: targeting.showSupportMessaging,
-                alreadyVisitedCount: targeting.alreadyVisitedCount,
-                countryCode: targeting.countryCode,
-                engagementBannerLastClosedAt: targeting.engagementBannerLastClosedAt,
-                subscriptionBannerLastClosedAt: targeting.subscriptionBannerLastClosedAt,
-                isPaidContent: targeting.isPaidContent,
-            };
-
-            res.send(response);
-        } catch (error) {
-            next(error);
+import { logError } from './utils/logging';
+import { buildEpicRouter } from './api/epicRouter';
+import { buildBannerRouter } from './api/bannerRouter';
+import { buildHeaderRouter } from './api/headerRouter';
+import { buildAmpEpicRouter } from './api/ampEpicRouter';
+import { buildModulesRouter } from './api/modulesRouter';
+
+const buildApp = (): Promise<Express> => {
+    const app = express();
+
+    app.use(express.json({ limit: '50mb' }));
+    app.use(compression());
+
+    const dotcomDevOrigins = ['http://localhost:3030', 'http://localhost:9000'];
+    const corsOrigin = () => {
+        switch (process.env.stage) {
+            case 'PROD':
+                return ['https://www.theguardian.com', ...dotcomDevOrigins];
+            case 'CODE':
+                return ['https://m.code.dev-theguardian.com', ...dotcomDevOrigins];
+            default:
+                return '*';
         }
-    },
-);
-
-app.post(
-    '/header',
-    bodyContainsAllFields(['tracking', 'targeting']),
-    async (req: express.Request, res: express.Response, next: express.NextFunction) => {
-        try {
-            const { tracking, targeting } = req.body;
-            const params = getQueryParams(req.query);
-            const response = await buildHeaderData(tracking, targeting, baseUrl(req), params, req);
-            res.send(response);
-        } catch (error) {
-            next(error);
-        }
-    },
-);
+    };
 
-/**
- * For component endpoints only. Data endpoints must never be cached.
- * Tell fastly to cache for 5 mins
- * Tell clients to cache for 2 mins
- */
-const setComponentCacheHeaders = (res: express.Response): void => {
-    if (isProd) {
-        res.setHeader('Surrogate-Control', 'max-age=300');
-        res.setHeader('Cache-Control', 'max-age=120');
+    const corsOptions = {
+        origin: corsOrigin(),
+    };
+    app.use(cors(corsOptions));
+    app.use(loggingMiddleware);
+    app.use(bodyParser.urlencoded({ extended: true }));
+
+    // Build the routers
+    app.use(buildEpicRouter());
```

Comment:
I really like this pattern (composing several sub routers to build up the overall app). I'm mostly thinking out loud here, but one tweak to it I've thought about before is for the individual routers to not have to know where they lie in the routing structure overall, and to pull that knowledge up to the app (i.e. here). So instead of prefixing routes under `buildAmpEpicRouter` with `/amp`, that would happen here like this:

```js
app.use(`/amp`, buildAmpEpicRouter())
```

I like this for a couple of reasons:

1. It's easier to see at a high level how the routes in the app are organized. (One thing I miss from Rails is that the routes are defined separately from controllers/handlers, so it's easy to get an overview from a single file of the routing structure of the app).
2. It's easier to make changes to the routing structure, as now that prefix only exists in one place.

What do you think?

Reply:
That makes sense.
Sadly amp is the only case in this app where we use a path prefix, so I'll just do it for amp!